server_windows.go

```diff
@@ -37,3 +44,76 @@ func (s *Server) toLocalPath(p string) string {
 
 	return lp
 }
+
+var kernel32, _ = syscall.LoadLibrary("kernel32.dll")
+var getLogicalDrivesHandle, _ = syscall.GetProcAddress(kernel32, "GetLogicalDrives")
+
+func bitsToDrives(bitmap uint32) []string {
+	var drive rune = 'A'
+	var drives []string
+
+	for bitmap != 0 {
+		if bitmap&1 == 1 {
+			drives = append(drives, string(drive))
+		}
+		drive++
+		bitmap >>= 1
+	}
+
+	return drives
+}
+
+func getDrives() ([]string, error) {
+	if ret, _, callErr := syscall.Syscall(uintptr(getLogicalDrivesHandle), 0, 0, 0, 0); callErr != 0 {
+		return nil, fmt.Errorf("GetLogicalDrives: %w", callErr)
+	} else {
+		drives := bitsToDrives(uint32(ret))
+		return drives, nil
+	}
+}
+
+type dummyDriveStat struct {
+	name string
+}
+
+func (s *dummyDriveStat) Name() string {
+	return s.name
+}
+func (s *dummyDriveStat) Size() int64 {
+	return 1024
+}
+func (s *dummyDriveStat) Mode() os.FileMode {
+	return os.FileMode(0755)
+}
+func (s *dummyDriveStat) ModTime() time.Time {
+	return time.Now()
+}
+func (s *dummyDriveStat) IsDir() bool {
+	return true
+}
+func (s *dummyDriveStat) Sys() any {
+	return nil
+}
+
+type WinRoot struct {
+	dummyFile
+}
+
+func (f *WinRoot) Readdir(int) ([]os.FileInfo, error) {
+	drives, err := getDrives()
+	if err != nil {
+		return nil, err
+	}
+	infos := []os.FileInfo{}
+	for _, drive := range drives {
+		infos = append(infos, &dummyDriveStat{drive})
+	}
+	return infos, nil
+}
+
+func openFileLike(path string, flag int, mode fs.FileMode) (FileLike, error) {
+	if path == "/" {
```

Comment:
This is a bug, which means we would never actually be returning `&WinRoot{}`.

`Server.toLocalPath` has already mangled `path` with at minimum `filepath.FromSlash` meaning this path should be now in local Windows format.

I think I would also like some sort of consideration that `\` is a valid path in Windows meaning “top-level of current drive” vs what we’re trying to do here, which is list the logical drives available.

We could maybe internally use here the filepath `\\.\` which is the start of the Win32 Device Namespace, and otherwise invalid.

Also, nothing here so far is going to force `/` to be followed by a volume indicator. So `/c:/windows` and `/windows` is likely going to point to the same place.

If we use the `\\.\` device namespace prefix, then at least we could distinguish `/c:/windows` as `\\.\c:\windows` from `/windows` as `\windows`. Otherwise, we’re going to need some hefty work here making sure we don’t just make the same weird situation where some paths are valid and can be opened despite being unlistable.

Reply:
So we'd also have to adjust `toLocalPath` to always use `\\.\`? And then automatically paths like `/windows` would resolve to `\\.\windows\` -> not valid anymore (which is wanted)?